module_utils/network_lsr/nm/active_connection.py

```diff
@@ -0,0 +1,127 @@
+        logging.debug(
+            "Deactiving {id} with timeout {timeout}".format(
+                id=nm_ac.get_id(), timeout=timeout
+            )
+        )
+        user_data = main_loop
+        handler_id = nm_ac.connect(
+            NM_AC_STATE_CHANGED_SIGNAL, _nm_ac_state_change_callback, user_data
+        )
+        logging.debug(
+            "Registered {signal} on NM.ActiveConnection {id}".format(
+                signal=NM_AC_STATE_CHANGED_SIGNAL, id=nm_ac.get_id()
+            )
+        )
+        if nm_ac.props.state != NM.ActiveConnectionState.DEACTIVATING:
+            nm_client = get_client()
+            user_data = (main_loop, nm_ac, handler_id)
+            nm_client.deactivate_connection_async(
+                nm_ac,
+                main_loop.cancellable,
+                _nm_ac_deactivate_call_back,
+                user_data,
+            )
+            logging.debug("Deactivating NM.ActiveConnection {0}".format(nm_ac.get_id()))
+        main_loop.run()
+    return True
+
+
+def _nm_ac_state_change_callback(nm_ac, state, reason, main_loop):
+    if main_loop.is_cancelled:
+        return
+    logging.debug(
+        "Got NM.ActiveConnection state change: {id}: {state} {reason}".format(
+            id=nm_ac.get_id(), state=state, reason=reason
+        )
+    )
+    if nm_ac.props.state == NM.ActiveConnectionState.DEACTIVATED:
+        logging.debug("NM.ActiveConnection {0} is deactivated".format(nm_ac.get_id()))
+        main_loop.quit()
+
+
+def _nm_ac_deactivate_call_back(nm_client, result, user_data):
+    main_loop, nm_ac, handler_id = user_data
+    logging.debug("NM.ActiveConnection deactivating callback")
+    if main_loop.is_cancelled:
+        if nm_ac:
+            nm_ac.handler_disconnect(handler_id)
+        return
+
+    try:
+        success = nm_client.deactivate_connection_finish(result)
+    except GLib.Error as e:
+        if e.matches(NM.ManagerError.quark(), NM.ManagerError.CONNECTIONNOTACTIVE):
+            success = True
+            logging.info(
+                "Connection is not active on {0}, no need to deactivate".format(
+                    nm_ac.get_id()
+                )
+            )
+            if nm_ac:
+                nm_ac.handler_disconnect(handler_id)
+            main_loop.quit()
+        else:
+            _deactivate_fail(
+                main_loop,
+                handler_id,
+                nm_ac,
+                "Failed to deactivate connection {id}, error={error}".format(
+                    id=nm_ac.get_id(), error=e
+                ),
+            )
+    except Exception as e:
+        _deactivate_fail(
+            main_loop,
+            handler_id,
+            nm_ac,
+            "Failed to deactivate connection {id}, error={error}".format(
+                id=nm_ac.get_id(), error=e
+            ),
+        )
+
+    if success:
+        logging.debug("NM.ActiveConnection {0} deactivated".format(nm_ac.get_id()))
```

Comment:
when the callback returns, in fact NetworkManager makes an effort to ensure that the active connection already changed state to DEACTIVATING state (or later). But it seems unnecessary to rely on that, and the previous code didn't do that. Previously, the `deactivate_connection_async()` only initiated the deactivation, and then the code would check whether the DEACTIVATED state was reached (with a timeout). Maybe it would be more robust, if a success from `deactivate_connection_finish()` does not yet automatically assume that the DEACTIVATED state was reached.

Reply:
Yes. That's correct. I have removed the `main_loop.quit()` when `success` returned by `deactivate_connection_finish()` and wait state change signal callback.